### PR TITLE
Allow SqliteAccountInfo to use different profile files

### PR DESCRIPTION
**Pull request by agoncharov-reef**

### CODE REVIEW

File: b2sdk/account_info/sqlite_account_info.py [hunk 1]
```diff
@@ -62,21 +70,7 @@ def __init__(self, file_name=None, last_upgrade_to_run=None):
...
-        elif XDG_CONFIG_HOME_ENV_VAR in os.environ:
-            config_home = os.environ[XDG_CONFIG_HOME_ENV_VAR]
-            user_account_info_path = os.path.join(config_home, 'b2', 'account_info')
-            if not os.path.exists(os.path.join(config_home, 'b2')):
-                os.makedirs(os.path.join(config_home, 'b2'), mode=0o755)
-        else:
-            user_account_info_path = B2_ACCOUNT_INFO_DEFAULT_FILE
-
-        self.filename = os.path.expanduser(user_account_info_path)
+        self.filename = self.get_user_account_info_path(file_name=file_name, profile=profile)
```
Comment (ppolewicz) [inline b2sdk/account_info/sqlite_account_info.py:73]:
```suggestion
        self.filename = self._get_user_account_info_path(file_name=file_name, profile=profile)
```

File: b2sdk/account_info/sqlite_account_info.py [hunk 2]
```diff
@@ -39,7 +42,7 @@ class SqliteAccountInfo(UrlPoolAccountInfo):
     completed.
     """
 
-    def __init__(self, file_name=None, last_upgrade_to_run=None):
+    def __init__(self, file_name=None, last_upgrade_to_run=None, profile=None):
```
Comment (ppolewicz) [inline b2sdk/account_info/sqlite_account_info.py:45]:
```suggestion
    def __init__(self, file_name=None, last_upgrade_to_run=None, profile: Optional[str] = None):
```

File: b2sdk/account_info/sqlite_account_info.py [hunk 3]
```diff
@@ -90,10 +84,42 @@ def __init__(self, file_name=None, last_upgrade_to_run=None):
             **dict(
                 B2_ACCOUNT_INFO_ENV_VAR=B2_ACCOUNT_INFO_ENV_VAR,
                 B2_ACCOUNT_INFO_DEFAULT_FILE=B2_ACCOUNT_INFO_DEFAULT_FILE,
+                B2_ACCOUNT_INFO_PROFILE_FILE=B2_ACCOUNT_INFO_PROFILE_FILE,
                 XDG_CONFIG_HOME_ENV_VAR=XDG_CONFIG_HOME_ENV_VAR,
             )
         )
 
+    @classmethod
+    def get_user_account_info_path(cls, file_name=None, profile=None):
```
Comment (ppolewicz) [inline b2sdk/account_info/sqlite_account_info.py:93]:
```suggestion
    def _get_user_account_info_path(cls, file_name=None, profile=None):
```

File: test/unit/account_info/test_sqlite_account_info.py [hunk 4]
```diff
@@ -55,3 +64,66 @@ def test_migrate_to_4(self):
...
+    @pytest.fixture(autouse=True)
+    def setup(self, monkeypatch):
+        monkeypatch.delenv(B2_ACCOUNT_INFO_ENV_VAR, raising=False)
+        monkeypatch.delenv(XDG_CONFIG_HOME_ENV_VAR, raising=False)
+        if os.path.exists(os.path.expanduser(B2_ACCOUNT_INFO_DEFAULT_FILE)):
+            os.remove(os.path.expanduser(B2_ACCOUNT_INFO_DEFAULT_FILE))
+
+    def test_invalid_profile_name(self):
+        with pytest.raises(ValueError):
+            SqliteAccountInfo.get_user_account_info_path(profile='&@(*$')
```
Comment (ppolewicz) [inline test/unit/account_info/test_sqlite_account_info.py:79]:
```suggestion
            SqliteAccountInfo._get_user_account_info_path(profile='&@(*$')
```

File: test/unit/account_info/test_sqlite_account_info.py [hunk 5]
```diff
@@ -55,3 +64,66 @@ def test_migrate_to_4(self):
...
+
+    def test_profile_and_file_name_conflict(self):
+        with pytest.raises(ValueError):
+            SqliteAccountInfo.get_user_account_info_path(file_name='foo', profile='bar')
```
Comment (ppolewicz) [inline test/unit/account_info/test_sqlite_account_info.py:83]:
```suggestion
            SqliteAccountInfo._get_user_account_info_path(file_name='foo', profile='bar')
```

File: test/unit/account_info/test_sqlite_account_info.py [hunk 6]
```diff
@@ -55,3 +64,66 @@ def test_migrate_to_4(self):
...
+
+    def test_profile_and_env_var_conflict(self, monkeypatch):
+        monkeypatch.setenv(B2_ACCOUNT_INFO_ENV_VAR, 'foo')
+        with pytest.raises(ValueError):
+            SqliteAccountInfo.get_user_account_info_path(profile='bar')
```
Comment (ppolewicz) [inline test/unit/account_info/test_sqlite_account_info.py:88]:
```suggestion
            SqliteAccountInfo._get_user_account_info_path(profile='bar')
```

File: test/unit/account_info/test_sqlite_account_info.py [hunk 7]
```diff
@@ -55,3 +64,66 @@ def test_migrate_to_4(self):
...
+
+    def test_profile_and_xdg_config_env_var(self, monkeypatch):
+        monkeypatch.setenv(XDG_CONFIG_HOME_ENV_VAR, os.path.join('~', 'custom'))
+        account_info_path = SqliteAccountInfo.get_user_account_info_path(profile='secondary')
```
Comment (ppolewicz) [inline test/unit/account_info/test_sqlite_account_info.py:92]:
```suggestion
        account_info_path = SqliteAccountInfo._get_user_account_info_path(profile='secondary')
```

File: test/unit/account_info/test_sqlite_account_info.py [hunk 8]
```diff
@@ -55,3 +64,66 @@ def test_migrate_to_4(self):
...
+        assert account_info_path == os.path.expanduser(
+            os.path.join('~', 'custom', 'b2', 'db-secondary.sqlite')
+        )
+
+    def test_profile(self):
+        account_info_path = SqliteAccountInfo.get_user_account_info_path(profile='foo')
```
Comment (ppolewicz) [inline test/unit/account_info/test_sqlite_account_info.py:98]:
```suggestion
        account_info_path = SqliteAccountInfo._get_user_account_info_path(profile='foo')
```

File: test/unit/account_info/test_sqlite_account_info.py [hunk 9]
```diff
@@ -55,3 +64,66 @@ def test_migrate_to_4(self):
...
+        assert account_info_path == os.path.expanduser(os.path.join('~', '.b2db-foo.sqlite'))
+
+    def test_file_name(self):
+        account_info_path = SqliteAccountInfo.get_user_account_info_path(
```
Comment (ppolewicz) [inline test/unit/account_info/test_sqlite_account_info.py:102]:
```suggestion
        account_info_path = SqliteAccountInfo._get_user_account_info_path(
```

File: test/unit/account_info/test_sqlite_account_info.py [hunk 10]
```diff
@@ -55,3 +64,66 @@ def test_migrate_to_4(self):
...
+            file_name=os.path.join('~', 'foo')
+        )
+        assert account_info_path == os.path.expanduser(os.path.join('~', 'foo'))
+
+    def test_env_var(self, monkeypatch):
+        monkeypatch.setenv(B2_ACCOUNT_INFO_ENV_VAR, os.path.join('~', 'foo'))
+        account_info_path = SqliteAccountInfo.get_user_account_info_path()
```
Comment (ppolewicz) [inline test/unit/account_info/test_sqlite_account_info.py:109]:
```suggestion
        account_info_path = SqliteAccountInfo._get_user_account_info_path()
```

File: test/unit/account_info/test_sqlite_account_info.py [hunk 11]
```diff
@@ -55,3 +64,66 @@ def test_migrate_to_4(self):
...
+        assert account_info_path == os.path.expanduser(os.path.join('~', 'foo'))
+
+    def test_default_file_if_exists(self, monkeypatch):
+        # ensure that XDG_CONFIG_HOME_ENV_VAR doesn't matter if default file exists
+        monkeypatch.setenv(XDG_CONFIG_HOME_ENV_VAR, 'some')
+        with open(os.path.expanduser(B2_ACCOUNT_INFO_DEFAULT_FILE), 'w') as account_file:
+            account_file.write('')
+        account_info_path = SqliteAccountInfo.get_user_account_info_path()
```
Comment (ppolewicz) [inline test/unit/account_info/test_sqlite_account_info.py:117]:
```suggestion
        account_info_path = SqliteAccountInfo._get_user_account_info_path()
```

File: test/unit/account_info/test_sqlite_account_info.py [hunk 12]
```diff
@@ -55,3 +64,66 @@ def test_migrate_to_4(self):
...
+        assert account_info_path == os.path.expanduser(B2_ACCOUNT_INFO_DEFAULT_FILE)
+
+    def test_xdg_config_env_var(self, monkeypatch):
+        monkeypatch.setenv(XDG_CONFIG_HOME_ENV_VAR, os.path.join('~', 'custom'))
+        account_info_path = SqliteAccountInfo.get_user_account_info_path()
```
Comment (ppolewicz) [inline test/unit/account_info/test_sqlite_account_info.py:122]:
```suggestion
        account_info_path = SqliteAccountInfo._get_user_account_info_path()
```

File: test/unit/account_info/test_sqlite_account_info.py [hunk 13]
```diff
@@ -55,3 +64,66 @@ def test_migrate_to_4(self):
...
+        assert account_info_path == os.path.expanduser(
+            os.path.join('~', 'custom', 'b2', 'account_info')
+        )
+
+    def test_default_file(self):
+        account_info_path = SqliteAccountInfo.get_user_account_info_path()
```
Comment (ppolewicz) [inline test/unit/account_info/test_sqlite_account_info.py:128]:
```suggestion
        account_info_path = SqliteAccountInfo._get_user_account_info_path()
```

File: test/unit/account_info/test_sqlite_account_info.py [hunk 14]
```diff
@@ -55,3 +64,66 @@ def test_migrate_to_4(self):
...
         assert (100, 5000000) == sizes
+
+
+class TestSqliteAccountProfileFileLocation:
+    @pytest.fixture(autouse=True)
+    def setup(self, monkeypatch):
+        monkeypatch.delenv(B2_ACCOUNT_INFO_ENV_VAR, raising=False)
+        monkeypatch.delenv(XDG_CONFIG_HOME_ENV_VAR, raising=False)
+        if os.path.exists(os.path.expanduser(B2_ACCOUNT_INFO_DEFAULT_FILE)):
+            os.remove(os.path.expanduser(B2_ACCOUNT_INFO_DEFAULT_FILE))
```
Comment (ppolewicz) [inline test/unit/account_info/test_sqlite_account_info.py:75]: That would delete default accountinfo on my development environment. Maybe there is a better way?

File: b2sdk/account_info/sqlite_account_info.py (hunk 2 above)
Comment (agoncharov-reef) [reply b2sdk/account_info/sqlite_account_info.py]: I just follow the convention to not use type hints in b2sdk code :)

**Pull request merged**